python/ray/monitor.py

```diff
@@ -0,0 +1,186 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import binascii
+from collections import Counter
+import logging
+import redis
+import time
+
+from ray.services import get_ip_address
+from ray.services import get_port
+
+# These variables must be kept in sync with the C codebase.
+# common/common.h
+DBClientID_SIZE = 20
+NIL_ID = b"\xff" * DBClientID_SIZE
+# common/task.h
+TASK_STATUS_LOST = 32
+# common/redis_module/ray_redis_module.c
+TASK_PREFIX = "TT:"
+DB_CLIENT_PREFIX = "CL:"
+DB_CLIENT_TABLE_NAME = b"db_clients"
+# local_scheduler/local_scheduler.h
+LOCAL_SCHEDULER_HEARTBEAT_TIMEOUT_MILLISECONDS = 100
+LOCAL_SCHEDULER_CLIENT_TYPE = b"local_scheduler"
+
+# Set up logging.
+logging.basicConfig()
+log = logging.getLogger()
+
+class Monitor(object):
+  """A monitor for Ray processes.
+
+  The monitor is in charge of cleaning up the tables in the global state after
+  processes have died. The monitor is currently not responsible for detecting
+  component failures.
+
+  Attributes:
+    redis: A connection to the Redis server.
+    subscribe_client: A pubsub client for the Redis server. This is used to
+      receive notifications about failed components.
+    local_schedulers: A set of the local scheduler IDs of all of the currently
+      live local schedulers in the cluster. In addition, this also includes
+      NIL_ID.
+  """
+  def __init__(self, redis_address, redis_port):
+    self.redis = redis.StrictRedis(host=redis_address, port=redis_port, db=0)
+    self.subscribe_client = self.redis.pubsub()
+
+    # Initialize data structures to keep track of the active database clients.
+    self.local_schedulers = set()
+    # Add the NIL_ID so that we don't accidentally mark tasks that aren't
+    # associated with a node as LOST during cleanup.
+    self.local_schedulers.add(NIL_ID)
+
+  def subscribe(self):
```

Comment:
Methods like this should eventually be collected into a python "state" module (I started doing this in experimental/state); I hope once we have full flatbuffer coverage we can use the python flatbuffer module to generate methods that read this state automatically (and get rid of code duplication like DBClientID_SIZE = 20)